invoices/settlement_interceptor.go

```diff
@@ -0,0 +1,244 @@
+// InterceptClientResponse is the response that is sent from the client during
+// an interceptor session. The response contains the settlement resolution for
+// the invoice.
+type InterceptClientResponse struct {
+	// HtlcResolution is the resolution that the client has chosen for the
+	// invoice.
+	HtlcResolution HtlcResolution
+
+	// InvoiceUpdateDesc is an optional update description that the client
+	// can provide to update the invoice.
+	InvoiceUpdateDesc fn.Option[InvoiceUpdateDesc]
+}
+
+// InterceptSession creates a session that is returned to the caller when an
+// invoice is submitted to this service. This session allows the caller to block
+// until the invoice is processed.
+type InterceptSession struct {
+	// ExitHtlcCircuitKey is the circuit key that identifies the HTLC which
+	// is involved in the invoice settlement.
+	ExitHtlcCircuitKey CircuitKey
+
+	// Invoice is the invoice that is being intercepted.
+	Invoice Invoice
+
+	// ClientResponseChannel is a channel that is populated with the
+	// client's interceptor response during an interceptor session.
+	ClientResponseChannel chan InterceptClientResponse
+
+	// Quit is a channel that is closed when the session is no longer
+	// needed.
+	Quit chan struct{}
+}
+
+// SettlementInterceptor is a service that intercepts invoices during the
+// settlement phase, enabling a subscribed client to determine the settlement
+// outcome.
+type SettlementInterceptor struct {
+	wg sync.WaitGroup
+
+	// callback is a client defined function that is called when an invoice
+	// is intercepted. This function gives the client the ability to
+	// determine the settlement outcome.
+	clientCallback SafeCallback
+
+	// activeSessions is a map of active intercept sessions that are used to
+	// manage the client query/response for a given invoice payment hash.
+	activeSessions lnutils.SyncMap[lntypes.Hash, InterceptSession]
+}
+
+// NewSettlementInterceptor creates a new SettlementInterceptor.
+func NewSettlementInterceptor() *SettlementInterceptor {
+	return &SettlementInterceptor{
+		activeSessions: lnutils.SyncMap[
+			lntypes.Hash, InterceptSession,
+		]{},
+	}
+}
+
+// HandleInvoice adds an invoice to the service's unsettled invoice queue. These
+// invoices will be forwarded to any subscribed client for a settlement
+// decision.
+func (s *SettlementInterceptor) HandleInvoice(invoice Invoice,
+	settlementHtlcCircuitKey CircuitKey) fn.Option[InterceptSession] {
+
+	// If there is no client callback set we will not handle the invoice
+	// further.
+	if !s.clientCallback.IsSet() {
+		return fn.None[InterceptSession]()
+	}
+
+	// Create and store a new intercept session for the invoice. We will use
+	// the payment hash as the storage/retrieval key for the session.
+	paymentHash := invoice.Terms.PaymentPreimage.Hash()
+	session := InterceptSession{
+		ExitHtlcCircuitKey:    settlementHtlcCircuitKey,
+		Invoice:               invoice,
+		ClientResponseChannel: make(chan InterceptClientResponse, 5),
```

Comment:
nit: Why is the buffer size 5 here? AFAICT we only read from this channel once later on.

Reply:
+1, also if there **is** the need for a specific value, rather wrap it around a const var and give it a good name :+1: